dev/service-registry/service-locator/src/main/scala/com/lightbend/lagom/gateway/AkkaHttpServiceGateway.scala

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <https://www.lightbend.com>
+ */
+package com.lightbend.lagom.gateway
+
+import java.net.InetSocketAddress
+import java.util.Locale
+import javax.inject.{ Inject, Named }
+
+import akka.actor.{ ActorRef, ActorSystem }
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.model.ws._
+import akka.pattern.ask
+import akka.stream.Materializer
+import akka.stream.scaladsl.{ Flow, Keep, Sink, Source }
+import akka.util.Timeout
+import com.lightbend.lagom.discovery.ServiceRegistryActor.{ Found, NotFound, Route, RouteResult }
+import com.lightbend.lagom.internal.javadsl.registry.ServiceRegistryService
+import org.slf4j.LoggerFactory
+import play.api.inject.ApplicationLifecycle
+import play.api.routing.Router.Routes
+import play.api.routing.SimpleRouter
+
+import scala.collection.immutable
+import scala.concurrent.{ Await, Future }
+import scala.concurrent.duration._
+
+class AkkaHttpServiceGatewayFactory @Inject() (lifecycle: ApplicationLifecycle, config: ServiceGatewayConfig,
+                                               @Named("serviceRegistryActor") registry: ActorRef)(implicit actorSystem: ActorSystem, mat: Materializer) {
+
+  def start(): InetSocketAddress = {
+    new AkkaHttpServiceGateway(lifecycle, config, registry).address
+  }
+}
+
+class AkkaHttpServiceGateway(lifecycle: ApplicationLifecycle, config: ServiceGatewayConfig, registry: ActorRef)(implicit actorSystem: ActorSystem, mat: Materializer) {
+
+  private val log = LoggerFactory.getLogger(classOf[AkkaHttpServiceGateway])
+
+  import actorSystem.dispatcher
+  private implicit val timeout = Timeout(5.seconds)
+  val http = Http()
+
+  private val handler = Flow[HttpRequest].mapAsync(4) { request =>
```

Comment:
The flow will be materialized once for every TCP connection, so this would then potentially allow concurrent processing of 4 requests to be inflight per connection (if the client does pipelining). `mapAsync` guarantees the emitted values are ordered as they came in though, so if request 1 is long, it will still cause head of line blocking of emitting the other responses (and I think this is also a requirement of pipelining, responses must come in the same order as the requests). Relevant doc section: http://doc.akka.io/docs/akka-http/10.0.9/scala/http/low-level-server-side-api.html#controlling-server-parallelism

I didn't read the rest of the logic, but non idempotent operations (POST?) should not be pipelined, and some other operations (DELETE, PUT?) should not if they touch the same resource as other of the pipelined requests as the ordering of the changes will be indeterministic.

I'm a bit rusty on how http pipelining works